### PR TITLE
Align test names, add feature detection and add per-URL verification.

### PR DESCRIPTION
**Pull request by yoavweiss**
Various improvements to the existing set of preload tests:

* Align the test names with the rest of the tests, i.e. use '-' instead of '_'.
* Add feature detection to make sure the tests consistently fail in non-supporting browsers.
* Make sure that download verification is based on the URL's existence in ResourceTiming, rather on the overall number of resources downloaded, which can be misleading.

/cc @jgraham

### CODE REVIEW

**Comment by w3c-bots** [comment]
# Chrome (unstable channel) #
Testing web-platform-tests at revision 1a85e61c7dd0f42a6639fdff0bd8aebe6b4e81dd
Using browser at version 58.0.3000.4 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/link-header-preload.html">/preload/link-header-preload.html</a></summary>


|                     Subtest                      | Results |
|--------------------------------------------------|---------|
|                                                  | OK      |
| `Makes sure that Link headers preload resources` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/link-header-preload-delay-onload.html">/preload/link-header-preload-delay-onload.html</a></summary>


|                                              Subtest                                              | Results |
|---------------------------------------------------------------------------------------------------|---------|
|                                                                                                   | OK      |
| `Makes sure that Link headers preload resources and block window.onload after resource discovery` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/preload-csp.sub.html">/preload/preload-csp.sub.html</a></summary>


|                    Subtest                     | Results |
|------------------------------------------------|---------|
|                                                | OK      |
| `Makes sure that preload requests respect CSP` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/delaying-onload-link-preload-after-discovery.html">/preload/delaying-onload-link-preload-after-discovery.html</a></summary>


|                                      Subtest                                      | Results |
|-----------------------------------------------------------------------------------|---------|
|                                                                                   | OK      |
| `Makes sure link preload preloaded resources are delaying onload after discovery` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/download-resources.html">/preload/download-resources.html</a></summary>


|                       Subtest                        | Results |
|------------------------------------------------------|---------|
|                                                      | OK      |
| `Makes sure that preloaded resources are downloaded` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/single-download-preload.html">/preload/single-download-preload.html</a></summary>


|                                 Subtest                                  | Results |
|--------------------------------------------------------------------------|---------|
|                                                                          | OK      |
| `Makes sure that preloaded resources are not downloaded again when used` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/onerror-event.html">/preload/onerror-event.html</a></summary>


|                             Subtest                             | Results |
|-----------------------------------------------------------------|---------|
|                                                                 | OK      |
| `Makes sure that preloaded resources trigger the onerror event` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/preload-with-type.html">/preload/preload-with-type.html</a></summary>


|                                       Subtest                                        | Results |
|--------------------------------------------------------------------------------------|---------|
|                                                                                      | OK      |
| `Makes sure that preloaded resources with a type attribute trigger the onload event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/avoid-delaying-onload-link-preload.html">/preload/avoid-delaying-onload-link-preload.html</a></summary>


|                                Subtest                                | Results |
|-----------------------------------------------------------------------|---------|
|                                                                       | OK      |
| `Makes sure link preload preloaded resources are not delaying onload` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/onload-event.html">/preload/onload-event.html</a></summary>


|                            Subtest                             | Results |
|----------------------------------------------------------------|---------|
|                                                                | OK      |
| `Makes sure that preloaded resources trigger the onload event` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/single-download-late-used-preload.html">/preload/single-download-late-used-preload.html</a></summary>


|                             Subtest                             | Results |
|-----------------------------------------------------------------|---------|
|                                                                 | OK      |
| `Ensure preloaded resources are not downloaded again when used` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/dynamic-adding-preload.html">/preload/dynamic-adding-preload.html</a></summary>


|                                Subtest                                 | Results |
|------------------------------------------------------------------------|---------|
|                                                                        | OK      |
| `Makes sure that a dynamically added preloaded resource is downloaded` | PASS    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/preload-default-csp.sub.html">/preload/preload-default-csp.sub.html</a></summary>


|                    Subtest                     | Results |
|------------------------------------------------|---------|
|                                                | OK      |
| `Makes sure that preload requests respect CSP` | PASS    |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4752/preload/fetch-destination.https.html">/preload/fetch-destination.https.html</a></summary>


|           Subtest           | Results |
|-----------------------------|---------|
|                             | ERROR   |
| `Fetch destination preload` | FAIL    |

</details>

**Comment by w3c-bots** [comment]
# Firefox (nightly channel) #
Testing web-platform-tests at revision 1a85e61c7dd0f42a6639fdff0bd8aebe6b4e81dd
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/link-header-preload.html">/preload/link-header-preload.html</a></summary>


|                     Subtest                      | Results |
|--------------------------------------------------|---------|
|                                                  | OK      |
| `Makes sure that Link headers preload resources` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/link-header-preload-delay-onload.html">/preload/link-header-preload-delay-onload.html</a></summary>


|                                              Subtest                                              | Results |
|---------------------------------------------------------------------------------------------------|---------|
|                                                                                                   | OK      |
| `Makes sure that Link headers preload resources and block window.onload after resource discovery` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/preload-csp.sub.html">/preload/preload-csp.sub.html</a></summary>


|                    Subtest                     | Results |
|------------------------------------------------|---------|
|                                                | OK      |
| `Makes sure that preload requests respect CSP` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/delaying-onload-link-preload-after-discovery.html">/preload/delaying-onload-link-preload-after-discovery.html</a></summary>


|                                      Subtest                                      | Results |
|-----------------------------------------------------------------------------------|---------|
|                                                                                   | OK      |
| `Makes sure link preload preloaded resources are delaying onload after discovery` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/download-resources.html">/preload/download-resources.html</a></summary>


|                       Subtest                        | Results |
|------------------------------------------------------|---------|
|                                                      | OK      |
| `Makes sure that preloaded resources are downloaded` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/single-download-preload.html">/preload/single-download-preload.html</a></summary>


|                                 Subtest                                  | Results |
|--------------------------------------------------------------------------|---------|
|                                                                          | OK      |
| `Makes sure that preloaded resources are not downloaded again when used` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/onerror-event.html">/preload/onerror-event.html</a></summary>


|                             Subtest                             | Results |
|-----------------------------------------------------------------|---------|
|                                                                 | OK      |
| `Makes sure that preloaded resources trigger the onerror event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/preload-with-type.html">/preload/preload-with-type.html</a></summary>


|                                       Subtest                                        | Results |
|--------------------------------------------------------------------------------------|---------|
|                                                                                      | OK      |
| `Makes sure that preloaded resources with a type attribute trigger the onload event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/avoid-delaying-onload-link-preload.html">/preload/avoid-delaying-onload-link-preload.html</a></summary>


|                                Subtest                                | Results |
|-----------------------------------------------------------------------|---------|
|                                                                       | OK      |
| `Makes sure link preload preloaded resources are not delaying onload` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/onload-event.html">/preload/onload-event.html</a></summary>


|                            Subtest                             | Results |
|----------------------------------------------------------------|---------|
|                                                                | OK      |
| `Makes sure that preloaded resources trigger the onload event` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/single-download-late-used-preload.html">/preload/single-download-late-used-preload.html</a></summary>


|                             Subtest                             | Results |
|-----------------------------------------------------------------|---------|
|                                                                 | OK      |
| `Ensure preloaded resources are not downloaded again when used` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/dynamic-adding-preload.html">/preload/dynamic-adding-preload.html</a></summary>


|                                Subtest                                 | Results |
|------------------------------------------------------------------------|---------|
|                                                                        | OK      |
| `Makes sure that a dynamically added preloaded resource is downloaded` | FAIL    |

</details>
<details>
<summary><a href="http://w3c-test.org/submissions/4752/preload/preload-default-csp.sub.html">/preload/preload-default-csp.sub.html</a></summary>


|                    Subtest                     | Results |
|------------------------------------------------|---------|
|                                                | OK      |
| `Makes sure that preload requests respect CSP` | FAIL    |

</details>
<details>
<summary><a href="https://w3c-test.org/submissions/4752/preload/fetch-destination.https.html">/preload/fetch-destination.https.html</a></summary>


|           Subtest           | Results |
|-----------------------------|---------|
|                             | OK      |
| `Fetch destination preload` | FAIL    |

</details>

File: preload/resources/preload_helper.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+function verifyPreloadAndRTSupport()
+{
+    var link = window.document.createElement("link");
+    var preloadSupported = link.relList && link.relList.supports("preload");
```
Comment (zcorpan) [inline preload/resources/preload_helper.js]: Make this
```
assert_true(link.relList.supports("preload"), "preload not supported");
```

Comment (zcorpan) [inline preload/resources/preload_helper.js]: hmm or maybe that results in a less clear message if `relList` is not supported. OK to leave as is

File: preload/single-download-late-used-preload.html [hunk 2]
```diff
@@ -2,16 +2,18 @@
...
     var link = document.getElementsByTagName("link")[0]
     assert_equals(link.as, "image");
     link.addEventListener("load", () => {
-        assert_equals(performance.getEntriesByType("resource").length, 3)
+        verifyNumberOfDownloads("resources/square.png?pipe=trickle(d1)", 1);
         var img = document.createElement("img");
         img.src = "resources/square.png?pipe=trickle(d1)";
         img.onload = () => {
-            assert_equals(performance.getEntriesByType("resource").length, 3);
+            verifyPreloadAndRTSupport();
```
Comment (zcorpan) [inline preload/single-download-late-used-preload.html]: This can happen earlier (first thing in `link load`)

Comment (zcorpan) [inline preload/single-download-late-used-preload.html]: Alternatively, do the feature check inside `verifyNumberOfDownloads` so the tests don't need to worry about it.

**Comment by zcorpan** [comment]
w3c-test:mirror

File: preload/resources/preload_helper.js (hunk 1 above)
Comment (yoavweiss) [inline preload/resources/preload_helper.js]: ok

File: preload/single-download-late-used-preload.html (hunk 2 above)
Comment (yoavweiss) [inline preload/single-download-late-used-preload.html]: done

Comment (yoavweiss) [inline preload/single-download-late-used-preload.html]: I prefer to explicitly verify feature support in each test, as not all of them use `verifyNumberOfDownloads`

File: preload/single-download-preload.html [hunk 3]
```diff
@@ -42,9 +43,19 @@
...
+            verifyNumberOfDownloads("resources/dummy.css", 1);
+            verifyNumberOfDownloads("resources/square.png", 1);
+            verifyNumberOfDownloads("resources/square.png?background", 1);
+            verifyNumberOfDownloads("/fonts/CanvasTest.ttf", 1);
+            verifyNumberOfDownloads("resources/dummy.xml?foobar", 0);
+            verifyNumberOfDownloads("/media/foo.vtt", 1);
+            verifyNumberOfDownloads("resources/dummy.xml", 1);
+            // FIXME: We should verify for video and audio as well, but they seem to (flakily?) trigger multiple partial requests.
+            t.done();
+            }), 0);
```
Comment (zcorpan) [inline preload/single-download-preload.html]: Wrong indentation on this line

**Review by zcorpan — APPROVED** [review]
LGTM % 1 nit

File: preload/single-download-preload.html (hunk 3 above)
Comment (yoavweiss) [inline preload/single-download-preload.html]: Fixed

**Pull request merged**